### PR TITLE
Extended the Calculator Integer keyboard

### LINKED ISSUE #7818: Extend the integer keyboard still further
There is still more in the primes package we can add.  @anastasia-mbithe I am really pleased with the way you have been able to implement this keyboard.

a) Do we still need the prime key at the bottom.  I am not sure what it is adding now?
b) nth prime needs a tooltip, e.g. Gives the n-th prime.  For example nth_prime(1000)= 7919.  Also when you include the parameters, it is given as `n =`.  That is incorrect, it should be `x =`.
c) Change the function behind Factorize.  Currently it uses the function in DescTools.  I would prefer the one in primes, i.e. primes::prime_factors( )
d) Add a row giving gcd, scm, coprime.   The tooltips are as follows:
`gcd `Greatest common divisor, for example gcd(18,42) = 6
`scm` Smallest common multiple, for example scm(18,42) = 126 (= 7*18 & 3*42)
`coprime` Also called mutually prime, for example coprime(30,77) = TRUE. (30 = 2*3*5, 77 = 7*11)
e) Add phi (maybe on same row with gcd, scm, coprime) tooltip is Euler’s Totient Function. For example phi(12) = 4 (1, 2, 5, 7 are less than 12 and coprime)

Then also:
f) Move the 2nd row down to a block on its own. 
g) Add 6 keys labelled twin, cousin, sexy and then on another line third, triplets, k_tuple

The tooltips are: 
Twin primes, for example, twin(0,10) gives (3,5), & (5,7)
Cousin primes, for example cousin(0,20) gives (3,7) & (13,17)
Sexy primes (6 in Latin is sex!), for example sexy(0,40) gives (23,29) & (31,37)
Third cousin primes, for example third(0,100) gives (89,97)
Sexy prime triplets, for example triplets(0,100) gives (47, 53, 59)
k_tuple general formula for all these keys.  For example k_tuple(0, 10, c(0,2)) gives twin primes

The full function names are:
k_tuple(min, max, tuple)
twin_primes(min, max)
cousin_primes(min, max)
sexy_primes(min, max)
sexy_prime_triplets(min, max)
third_cousin_primes(min, max)

The code I used was, for example:
`rep(primes::third_cousin_primes(0,10000),len=1000)`

for a data frame on length 1000.  It should fill in the length automatically.  This is not ideal, but it works.  It repeats the sequences to the length of the data frame.
I would prefer to have the sequence just once and then use the `length` function to make the resulting variable the right length.  This is a bigger task.  Possibly it adds a length option in the Position button.

### PR DESCRIPTION
**Pull request by anastasia-mbithe**
Fixes #7818
This is still in progress. The last part of the issue is yet to be addressed (adding length to the position control).

### CODE REVIEW

**Comment by rdstern** [comment]
@anastasia-mbithe this all looks great.
a) Could you add the rep addition to the new prime, e.g. twin keys where needed - as explained in the issue.  You already added a similar feature in the prime key, to include the length of the data frame.
b) Add an s to the prime key, so it becomes `primes`.
c) Move the 5 prime keys from the main set to the top of the new set.  Have primes first, then n-th prime and next prime.  Then also is prime and coprime.

Many thanks

**Comment by anastasia-mbithe** [comment]
@rdstern , having the rep function for the primes only works for twin and triplets functions though on very limited occasions; when the length of the output is 1.
Its functionality depends on the length of the dataset and the length of the resulting output, so it mostly bugs because the length of the dataset is not divisible by the length of the resulting output from the function.
 
Even having: rep(primes::third_cousin_primes(0,10000),len=1000) in a dataset of 1000 rows still bugs because the length of the result is 101 which does not divide 1000.

**Comment by rdstern** [comment]
@anastasia-mbithe that's very odd, because it is just what the rep is supposed to do.

I tried with the rep in front of your functions.  I used the usual survey, so had 36 and it works fine for all your keys.  So, I am not sure what is going on?

For example - from my log file:

`calc6 <- rep(primes::sexy_prime_triplets(2 ,10000 ),len=36)
`
and:
`calc7 <- rep(primes::k_tuple( 2,10000 ,c(0,8)),len=36)`

image

**Comment by anastasia-mbithe** [comment]
@rdstern , I just added the rep function (am not sure why it was bugging yesterday, today the function works smoothly). I have also set the default to (0,100) for min and max since there has to be a default while implementing the function in this way. Also, am not sure if you want the parameters shown as well; for min, max or tuple.

**Review by rdstern — COMMENTED** [review]
@anastasia-mbithe the changes are very interesting.  Can you show them to @lilyclements .  You have included the specific values 0 and 100 for each of these special prime keys.  That's different to the other keys.  On the other hand it makes it very easy to adapt those values - so I rather like it.    It makes those keys easier to use, and I can't see it is doing any harm.  So I am inclined to suggest I should approve as it stands.  I'd just like Lily to give a second opinion.

**Review by lilyclements — APPROVED** [review]
Looks good to me!

**Review by rdstern — APPROVED** [review]

File: instat/ucrCalculator.vb [hunk 1]
```diff
@@ -3120,4 +3132,174 @@ Public Class ucrCalculator
...
+        End If
+    End Sub
+
+    Private Sub cmdTwin_Click(sender As Object, e As EventArgs) Handles cmdTwin.Click
+        Dim clsGetDataframeFunction As New RFunction
+        Dim clsDataFunction As New RFunction
+        Dim clsTwinPrimeFunction As New RFunction
+        Dim clsRepFunction As New RFunction
+
+        clsGetDataframeFunction.AddParameter("data_name", ucrSelectorForCalculations.ucrAvailableDataFrames.cboAvailableDataFrames.SelectedItem, bIncludeArgumentName:=False)
```
Comment (lloyddewit) [inline instat/ucrCalculator.vb:3174]: This function does not have a package name or R command. Is that correct?
(this comment also applies to the other functions below)

File: instat/ucrCalculator.vb [hunk 2]
```diff
@@ -3120,4 +3132,174 @@ Public Class ucrCalculator
...
+
+    Private Sub cmdPhi_Click(sender As Object, e As EventArgs) Handles cmdPhi.Click
+        If chkShowParameters.Checked Then
+            ucrReceiverForCalculation.AddToReceiverAtCursorPosition("primes::phi(n= )", 2)
+        Else
+            ucrReceiverForCalculation.AddToReceiverAtCursorPosition("primes::phi()", 1)
+        End If
+    End Sub
+
+    Private Sub cmdTwin_Click(sender As Object, e As EventArgs) Handles cmdTwin.Click
```
Comment (lloyddewit) [inline instat/ucrCalculator.vb:3168]: This sub seems to have the same structure as the 4 subs below it. The only thing that seems to vary is the name of the R command (e.g. "twin_primes").
Could we remove the duplication (e.g. add a private function that takes the name of the R command as a parameter)?

**Review by lloyddewit — CHANGES_REQUESTED** [review]

**Comment by anastasia-mbithe** [comment]
@lloyddewit, Kindly have a look at the changes made.

File: instat/ucrCalculator.vb (hunk 2 above)
Comment (lloyddewit) [reply instat/ucrCalculator.vb:3179]: Thank you for moving some of the duplicated code into the `CalculatorFunctions()` function.
Is there even more duplicated code we could move? For example:
```
    Private Sub CalculatorFunctions(strRCommand As String)
        Dim clsPrimesFunction As New RFunction

        clsPrimesFunction.SetPackageName("primes")
        clsPrimesFunction.SetRCommand(strRCommand)
        clsPrimesFunction.AddParameter("min", "0", iPosition:=0)
        clsPrimesFunction.AddParameter("max", "100", iPosition:=1)

        clsRepFunction.AddParameter("x", clsRFunctionParameter:=clsPrimesFunction, iPosition:=0, bIncludeArgumentName:=False)

        clsDataFunction.SetRCommand("nrow")
        clsDataFunction.AddParameter("x", ucrSelectorForCalculations.ucrAvailableDataFrames.cboAvailableDataFrames.SelectedItem, bIncludeArgumentName:=False)

        clsRepFunction.SetRCommand("rep")
        clsRepFunction.AddParameter("len", clsRFunctionParameter:=clsDataFunction, iPosition:=1)

        ucrReceiverForCalculation.AddToReceiverAtCursorPosition(clsRepFunction.ToScript, 0)
    End Sub

    Private Sub cmdTwin_Click(sender As Object, e As EventArgs) Handles cmdTwin.Click
        CalculatorFunctions("twin_primes")
    End Sub
```

**Comment by lloyddewit** [comment]
@anastasia-mbithe Thank you for the changes, if you can resolve the final open comment above, then I'm happy to approve

**Comment by N-thony** [comment]
> @lloyddewit, I resolved the pending issue. Kindly have a look at it.

@lloyddewit please have look

File: instat/ucrCalculator.vb [hunk 3]
```diff
@@ -3120,4 +3132,99 @@ Public Class ucrCalculator
...
+        clsDataFunction.SetRCommand("nrow")
+        clsDataFunction.AddParameter("x", ucrSelectorForCalculations.ucrAvailableDataFrames.cboAvailableDataFrames.SelectedItem, bIncludeArgumentName:=False)
+
+        clsRepFunction.SetRCommand("rep")
+        clsRepFunction.AddParameter("x", clsRFunctionParameter:=clsPrimesFunction, iPosition:=0, bIncludeArgumentName:=False)
+        clsRepFunction.AddParameter("len", clsRFunctionParameter:=clsDataFunction, iPosition:=1)
+
+        ucrReceiverForCalculation.AddToReceiverAtCursorPosition(clsRepFunction.ToScript, 0)
+    End Sub
+
```
Comment (N-thony) [inline instat/ucrCalculator.vb:3185]: one line space?

**Review by N-thony — CHANGES_REQUESTED** [review]
Great, just some minor comments

**Comment by anastasia-mbithe** [comment]
This is also complete as well. The changes requested are already resolved.

**Review by lloyddewit — APPROVED** [review]

**Comment by lloyddewit** [comment]
@anastasia-mbithe thank you for the changes. I'm happy to approve.

@N-thony thank you for the peer review comments. I checked and they all look resolved. If not, then please let us know.

@rdstern there have been non-functional changes since your last review. Please could you restest/approve? thanks

**Comment by N-thony** [comment]
> @anastasia-mbithe thank you for the changes. I'm happy to approve.
> 
> @N-thony thank you for the peer review comments. I checked and they all look resolved. If not, then please let us know.
> 
> @rdstern there have been non-functional changes since your last review. Please could you restest/approve? thanks

@lloyddewit looks good to me.

**Review by rdstern — APPROVED** [review]
Looks good.

**Pull request merged**